Port WebExtension Manifest Background and Inspector Content

ItsJamie9494 opened the pull request:
#### 728527cd9838203667cd707899aa4b87c6462175
<pre>
Port WebExtension Manifest Background and Inspector Content
<a href="https://webkit.org/b/280041">https://webkit.org/b/280041</a>

Reviewed by Timothy Hatcher.

Ports the WebExtension manifest background content and inspector content to C++.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::resourceStringForPath):
(WebKit::WebExtension::resourceDataForPath):
(WebKit::WebExtension::hasRequestedPermission const): Deleted.
(WebKit::WebExtension::hasBackgroundContent): Deleted.
(WebKit::WebExtension::backgroundContentIsPersistent): Deleted.
(WebKit::WebExtension::backgroundContentUsesModules): Deleted.
(WebKit::WebExtension::backgroundContentIsServiceWorker): Deleted.
(WebKit::WebExtension::backgroundContentPath): Deleted.
(WebKit::WebExtension::generatedBackgroundContent): Deleted.
(WebKit::WebExtension::populateBackgroundPropertiesIfNeeded): Deleted.
(WebKit::WebExtension::hasInspectorBackgroundPage): Deleted.
(WebKit::WebExtension::inspectorBackgroundPagePath): Deleted.
(WebKit::WebExtension::populateInspectorPropertiesIfNeeded): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
(WebKit::WebExtension::hasRequestedPermission const):
(WebKit::WebExtension::hasBackgroundContent):
(WebKit::WebExtension::backgroundContentIsPersistent):
(WebKit::WebExtension::backgroundContentUsesModules):
(WebKit::WebExtension::backgroundContentIsServiceWorker):
(WebKit::WebExtension::backgroundContentPath):
(WebKit::WebExtension::generatedBackgroundContent):
(WebKit::WebExtension::populateBackgroundPropertiesIfNeeded):
(WebKit::WebExtension::hasInspectorBackgroundPage):
(WebKit::WebExtension::inspectorBackgroundPagePath):
(WebKit::WebExtension::populateInspectorPropertiesIfNeeded):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:

Canonical link: <a href="https://commits.webkit.org/284078@main">https://commits.webkit.org/284078@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e7ed3c7bacd1dc34174ff9965d7e5f2d943a5049

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68393 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47785 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21052 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72460 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19538 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55581 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19354 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54607 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13012 "Passed tests") 
| [⏳ 🧪 webkitperl ](https://ews-build.webkit.org/#/builders/WebKitPerl-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43690 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59060 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35070 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40358 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17895 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62315 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16828 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74152 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12364 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16093 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62059 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12403 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59138 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62080 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10002 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3617 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10401 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43586 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/44660 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45854 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44402 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -480,10 +468,10 @@
     bool isServiceWorker = backgroundContentIsServiceWorker();
     if (!isServiceWorker && [path isEqualToString:generatedBackgroundPageFilename])
-        return generatedBackgroundContent();
+        return static_cast<NSString *>(generatedBackgroundContent());
```
Shouldn't need the cast, since the function return type acts as one here.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -480,10 +468,10 @@
 
     if (isServiceWorker && [path isEqualToString:generatedBackgroundServiceWorkerFilename])
-        return generatedBackgroundContent();
+        return static_cast<NSString *>(generatedBackgroundContent());
```
Ditto.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +220,270 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+
+    if (m_backgroundScriptPaths.size()) {
+        if (backgroundContentIsServiceWorker()) {
+            static const NeverDestroyed<String> backgroundContentString = generatedBackgroundServiceWorkerFilename;
+            return backgroundContentString;
```
Just return `generatedBackgroundServiceWorkerFilename` and don't use `NeverDestroyed` here.

ItsJamie9494 replied:
If we don't copy the string this way, it complains about returning a local temporary

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +220,270 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+            return backgroundContentString;
+        }
+
+        static const NeverDestroyed<String> backgroundContentString = generatedBackgroundPageFilename;
+        return backgroundContentString;
```
Ditto.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +220,270 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+    bool usesModules = backgroundContentUsesModules();
+
+    RefPtr scriptsArray = JSON::Array::create();
+    for (auto scriptPath : m_backgroundScriptPaths) {
```
```suggestion
    for (auto& scriptPath : m_backgroundScriptPaths) {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +220,270 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+    RefPtr scriptsArray = JSON::Array::create();
+    for (auto scriptPath : m_backgroundScriptPaths) {
+        if (!scriptPath)
+            continue;
```
This shouldn't be needed, since this is a `String`. And it gets filtered for null/empty earlier.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +220,270 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+        if (usesModules)
+            format.append(" type=\"module\" "_s);
+        format.append("src=\""_s, scriptPath, "\"></script>"_s);
+        scriptsArray->pushString(format.toString());
```
```suggestion
        format.append("src=\""_s, scriptPath, "\"></script>"_s);

        scriptsArray->pushString(format.toString());
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +220,270 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+    bool isServiceWorker = backgroundContentIsServiceWorker();
+    bool usesModules = backgroundContentUsesModules();
+
+    RefPtr scriptsArray = JSON::Array::create();
```
This should just be a `Vector<String>`.
```suggestion
    Vector<String> scripts;
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +220,270 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+            }
+
+            format.append("importScripts(\""_s, scriptPath, "\");"_s);
+            scriptsArray->pushString(format.toString());
```
```suggestion
            format.append("importScripts(\""_s, scriptPath, "\");"_s);

            scriptsArray->pushString(format.toString());
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +220,270 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+
+    for (Ref scriptPath : *scriptsArray) {
+        if (!scriptPath)
+            continue;
```
Will be simpler with the `Vector`.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +220,270 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+        m_backgroundScriptPaths = makeStringVector(*backgroundScriptPaths);
+    }
+
+    auto supportedEnvironments = HashSet<String>({ backgroundDocumentManifestKey, backgroundServiceWorkerManifestKey });
```
```suggestion
    HashSet<String> supportedEnvironments { backgroundDocumentManifestKey, backgroundServiceWorkerManifestKey };
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +220,270 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+        });
+
+        for (auto i = filteredEnvironments->begin(); i != filteredEnvironments->end(); ++i)
+            preferredEnvironments.add(i->get().asString());
```
```suggestion
        for (Ref environment : *filteredEnvironments)
            preferredEnvironments.add(environment->asString());
```

ItsJamie9494 replied:
for whatever reason, this fails if it doesn't manually use begin or end. I haven't figured out a way to make it not fail with an error similar to `indirection requires pointer operand ('WTF::Ref<WTF::JSONImpl::Array>' invalid)`

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +220,270 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+    } else if (backgroundManifestObject->getValue(backgroundPreferredEnvironmentManifestKey))
+        recordError(createError(Error::InvalidBackgroundContent, WEB_UI_STRING("Manifest `background` entry has an empty or invalid `preferred_environment` key.", "WKWebExtensionErrorInvalidBackgroundContent description for empty or invalid preferred environment key")));
+
+    for (String environment : preferredEnvironments) {
```
```suggestion
    for (auto& environment : preferredEnvironments) {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +220,270 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+    for (String environment : preferredEnvironments) {
+        if (environment == backgroundDocumentManifestKey) {
+            m_backgroundContentEnvironment = Environment::Document;
+            m_backgroundServiceWorkerPath = String();
```
```suggestion
            m_backgroundServiceWorkerPath = nullString();
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +220,270 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+
+            if (!m_backgroundPagePath.isEmpty()) {
+                // Page takes precedence over scripts and service worker.
+                m_backgroundScriptPaths = Vector<String>();
```
```suggestion
                m_backgroundScriptPaths = { };
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +220,270 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+
+        if (environment == backgroundServiceWorkerManifestKey) {
+            m_backgroundContentEnvironment = Environment::ServiceWorker;
+            m_backgroundPagePath = String();
```
```suggestion
            m_backgroundPagePath = nullString();
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +220,270 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+
+            if (!m_backgroundServiceWorkerPath.isEmpty()) {
+                // Page takes precedence over scripts and service worker.
+                m_backgroundScriptPaths = Vector<String>();
```
```suggestion
                m_backgroundScriptPaths = { };
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +220,270 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+    if (!preferredEnvironments.size()) {
+        // Page takes precedence over service worker.
+        if (!m_backgroundPagePath.isEmpty())
+            m_backgroundServiceWorkerPath = String();
```
```suggestion
            m_backgroundServiceWorkerPath = nullString();
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +220,270 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+        // Scripts takes precedence over page and service worker.
+        if (m_backgroundScriptPaths.size()) {
+            m_backgroundServiceWorkerPath = String();
+            m_backgroundPagePath = String();
```
```suggestion
            m_backgroundServiceWorkerPath = nullString();
            m_backgroundPagePath = nullString();
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +220,270 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+
+bool WebExtension::hasInspectorBackgroundPage()
+{
+    return !m_inspectorBackgroundPagePath.isEmpty();
```
```suggestion
    populateInspectorPropertiesIfNeeded();
    return !m_inspectorBackgroundPagePath.isEmpty();
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -90,6 +107,11 @@ double WebExtension::manifestVersion()
 
+bool WebExtension::hasRequestedPermission(String permission) const
+{
+    return m_permissions.contains(permission);
```
This is missing `populatePermissionsPropertiesIfNeeded()`, which I just realize was missing in the original impl!

ItsJamie9494 replied:
hasRequestedPermission is const, but populate* is not.

xeenon replied:
Hmm, try dropping `const`? Or do a dance with `mutable` or `const_cast`?

xeenon replied:
Or at minimum file a bug and add a FIXME comment so we don't forget to fix this. Since it isn't new, that is fine for now.

ItsJamie9494 replied:
https://bugs.webkit.org/show_bug.cgi?id=280101

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +220,264 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+
+    HashSet<String> supportedEnvironments = { backgroundDocumentManifestKey, backgroundServiceWorkerManifestKey };
+
+    HashSet<String> preferredEnvironments;
```
This needs to be a `Vector`, it was an `NSOrderedSet` before, because order matters.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +221,264 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+        m_backgroundScriptPaths = makeStringVector(*backgroundScriptPaths);
+    }
+
+    Vector<String> supportedEnvironments = { backgroundDocumentManifestKey, backgroundServiceWorkerManifestKey };
```
This could have stayed as `HashSet` since all we do with it is call `contains()`.

xeenon replied:
```suggestion
    HashSet<String> supportedEnvironments = { backgroundDocumentManifestKey, backgroundServiceWorkerManifestKey };
```

ItsJamie9494 commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +221,267 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+        if (supportedEnvironments.contains(environment))
+            preferredEnvironments.append(environment);
+    } else if (RefPtr environments = backgroundManifestObject->getArray(backgroundPreferredEnvironmentManifestKey)) {
+        Ref filteredEnvironments = filterObjects(*environments, [supportedEnvironments](auto& value) {
```
This here isn't working properly it seems, the BackgroundPreferredEnvironmentParsing test crashes here due to an assertion failure inside the HashSet. I'm not sure if it's not capturing the right values or what, but something with the HashSet in the lambda isn't working.

ItsJamie9494 replied:
Switching supportedEnvironments back to a Vector seems to work better here.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +221,267 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+        return nullString();
+
+    if (m_backgroundScriptPaths.size())
+        return nullString();
```
I think you have a bug here — leading to no tests running and timing out at 30 seconds each. This should be:

```suggestion
    if (m_backgroundScriptPaths.isEmpty())
        return nullString();
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +221,267 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+    if (!m_backgroundServiceWorkerPath.isEmpty())
+        return m_backgroundServiceWorkerPath;
+
+    if (m_backgroundScriptPaths.size()) {
```
```suggestion
    if (!m_backgroundScriptPaths.isEmpty()) {
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +221,267 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+bool WebExtension::hasBackgroundContent()
+{
+    populateBackgroundPropertiesIfNeeded();
+    return m_backgroundScriptPaths.size() || !m_backgroundPagePath.isEmpty() || !m_backgroundServiceWorkerPath.isEmpty();
```
```suggestion
    return !m_backgroundScriptPaths.isEmpty() || !m_backgroundPagePath.isEmpty() || !m_backgroundServiceWorkerPath.isEmpty();
```

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
```diff
@@ -198,6 +221,267 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+    if (!isServiceWorker)
+        generatedBackgroundContent.append("<!DOCTYPE html>\n<body>\n"_s);
+
+    for (auto scriptPath : scripts)
```
```suggestion
    for (auto& scriptPath : scripts)
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -478,12 +466,11 @@
-    if (!isServiceWorker && [path isEqualToString:generatedBackgroundPageFilename])
-        return generatedBackgroundContent();
+    if ([path isEqualToString:generatedBackgroundPageFilename] && !backgroundContentIsServiceWorker())
+        return [static_cast<NSString *>(generatedBackgroundContent()) dataUsingEncoding:NSUTF8StringEncoding];
```
Not a data result, this line should not change.
```suggestion
        return generatedBackgroundContent();
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -478,12 +466,11 @@
-    if (isServiceWorker && [path isEqualToString:generatedBackgroundServiceWorkerFilename])
-        return generatedBackgroundContent();
+    if ([path isEqualToString:generatedBackgroundServiceWorkerFilename] && !backgroundContentIsServiceWorker())
+        return [static_cast<NSString *>(generatedBackgroundContent()) dataUsingEncoding:NSUTF8StringEncoding];
```
Ditto.
```suggestion
        return generatedBackgroundContent();
```

xeenon replied:
Actually these two ifs, this can be simplified.

```suggestion
if ([path isEqualToString:generatedBackgroundPageFilename] || [path isEqualToString:generatedBackgroundServiceWorkerFilename])
    return generatedBackgroundContent();
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -478,11 +466,10 @@
-    bool isServiceWorker = backgroundContentIsServiceWorker();
-    if (!isServiceWorker && [path isEqualToString:generatedBackgroundPageFilename])
+    if ([path isEqualToString:generatedBackgroundPageFilename] && !backgroundContentIsServiceWorker())
         return generatedBackgroundContent();
```
This can be simplified now that I look at it:
```suggestion
    if ([path isEqualToString:generatedBackgroundPageFilename] || [path isEqualToString:generatedBackgroundServiceWorkerFilename])
        return generatedBackgroundContent();
```
Then delete the second `if`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -478,11 +466,10 @@
 
-    if (isServiceWorker && [path isEqualToString:generatedBackgroundServiceWorkerFilename])
+    if ([path isEqualToString:generatedBackgroundServiceWorkerFilename] && !backgroundContentIsServiceWorker())
         return generatedBackgroundContent();
```
```suggestion
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -548,12 +535,11 @@
-    if (!isServiceWorker && [path isEqualToString:generatedBackgroundPageFilename])
-        return [generatedBackgroundContent() dataUsingEncoding:NSUTF8StringEncoding];
+    if ([path isEqualToString:generatedBackgroundPageFilename] && !backgroundContentIsServiceWorker())
+        return [static_cast<NSString *>(generatedBackgroundContent()) dataUsingEncoding:NSUTF8StringEncoding];
```
Ditto.
```suggestion
    if ([path isEqualToString:generatedBackgroundPageFilename] || [path isEqualToString:generatedBackgroundServiceWorkerFilename])
        return [static_cast<NSString *>(generatedBackgroundContent()) dataUsingEncoding:NSUTF8StringEncoding];
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -548,12 +535,11 @@
-    if (isServiceWorker && [path isEqualToString:generatedBackgroundServiceWorkerFilename])
-        return [generatedBackgroundContent() dataUsingEncoding:NSUTF8StringEncoding];
+    if ([path isEqualToString:generatedBackgroundServiceWorkerFilename] && !backgroundContentIsServiceWorker())
+        return [static_cast<NSString *>(generatedBackgroundContent()) dataUsingEncoding:NSUTF8StringEncoding];
```
```suggestion
```

ItsJamie9494 committed:
Port WebExtension Manifest Background and Inspector Content
https://webkit.org/b/280041

Reviewed by Timothy Hatcher.

Ports the WebExtension manifest background content and inspector content to C++.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::resourceStringForPath):
(WebKit::WebExtension::resourceDataForPath):
(WebKit::WebExtension::hasRequestedPermission const): Deleted.
(WebKit::WebExtension::hasBackgroundContent): Deleted.
(WebKit::WebExtension::backgroundContentIsPersistent): Deleted.
(WebKit::WebExtension::backgroundContentUsesModules): Deleted.
(WebKit::WebExtension::backgroundContentIsServiceWorker): Deleted.
(WebKit::WebExtension::backgroundContentPath): Deleted.
(WebKit::WebExtension::generatedBackgroundContent): Deleted.
(WebKit::WebExtension::populateBackgroundPropertiesIfNeeded): Deleted.
(WebKit::WebExtension::hasInspectorBackgroundPage): Deleted.
(WebKit::WebExtension::inspectorBackgroundPagePath): Deleted.
(WebKit::WebExtension::populateInspectorPropertiesIfNeeded): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
(WebKit::WebExtension::hasRequestedPermission const):
(WebKit::WebExtension::hasBackgroundContent):
(WebKit::WebExtension::backgroundContentIsPersistent):
(WebKit::WebExtension::backgroundContentUsesModules):
(WebKit::WebExtension::backgroundContentIsServiceWorker):
(WebKit::WebExtension::backgroundContentPath):
(WebKit::WebExtension::generatedBackgroundContent):
(WebKit::WebExtension::populateBackgroundPropertiesIfNeeded):
(WebKit::WebExtension::hasInspectorBackgroundPage):
(WebKit::WebExtension::inspectorBackgroundPagePath):
(WebKit::WebExtension::populateInspectorPropertiesIfNeeded):
* Source/WebKit/UIProcess/Extensions/WebExtension.h:

Canonical link: https://commits.webkit.org/284078@main
